### ISSUE TITLE
Decide on Position/Input API

### ISSUE THREAD

**Issue by olafurpg** [scalacenter/macros#31]
Follow up from discussions in https://github.com/scalacenter/macros/pull/29/files#r144877334

* Position.start/end
* Input/SourceFile (should it exist?
* Input.contents (if Input is exposed)
* Position.lineContent

IMO, we should support all of them. These APIs are used (one example https://github.com/lihaoyi/sourcecode/blob/master/sourcecode/shared/src/main/scala/sourcecode/SourceContext.scala) and they seem to be easy to support by different backends.

**Comment by liufengyun** [comment]
Another question is, should `enclosingPosition` and `tree.pos` have the same type.

In the macro system, there is going to be two different usage of positions:

- `enclosingLocation` that eventually exposes line numbers, files to end Scala users via macro authors.
- `tree.pos`: so far, the only usage of this is for error reporting service in backend, which usually takes a position. It's not exposed to end Scala users, just for macro authors.

For the first one, it's a simple promise to be fulfilled by backends, and it's fine to have a case class for that in the macro system.

For the second usage, we don't need to expose any API on `Pos`, as there's no need to know what they are, thus it's fine to just have an abstract type for them.

If we don't want to give macro authors unnecessary capabilities, the separation seems to be justified. For backends, the implementation will also be easier.

**Comment by olafurpg** [comment]
> For the second usage, we don't need to expose any API on Pos, as there's no need to know what they are, thus it's fine to just have an abstract type for them.

What would be the migration story for macros like `utest.Tracer` that use `tree.pos.lineContent`?

https://github.com/lihaoyi/utest/blob/cdea6539e4915b27ecc295e9f73a9ba5b072c4d3/utest/shared/src/main/scala/utest/asserts/Tracer.scala#L72

**Comment by liufengyun** [comment]
If we limit `assert` to just one arguments, then the line can be retrieved from the source file by the enclosing line number.

There is a difference between the macro system providing `tree.pos.lineContent` and `enclosingLocation: Option[(LineNum, Path)]`. In the former, the macro system makes much more complex assumptions and promises to support more powerful API.

I was also thinking what semantics we should give to `enclosingLocation` when the macro itself is expanded from another macro. Position of trees is a tricky issue in the presence of macros and synthetic trees.

**Comment by olafurpg** [comment]
I agree that macros and synthetic trees complicate positions. The convention in scala.reflect is to use positions of original code to report errors on expanded code

```scala
[error]      value asset is not a member of object Predef
[error]      L11:    Asserts.assert(!a)
[error]      L11:                  ^
```
This results in cryptic error messages for macro authors and users.

To prevent this, we could maybe change the signature to return some form of `tree.pos: Either[PositionError, Position]`, where PositionError (name up for debate) is an ADT of common errors like

- NoPosition
- Synthetic(enclosing: Position), leave it to macro author to decide whether to use enclosing position or not.
- whatever else we can come up with to encode potential errors

> If we limit assert to just one arguments, then the line can be retrieved from the source file by the enclosing line number.

This is an assumption about the macro library. It may be a fine limitation in this case for assert, but it might be a blocker for other libraries. Following Occam's razor, we should try not to make too many assumptions about macro libraries, the same way we already do for compiler engines.
